### DISCUSSION TITLE
如何将options转为类似于config.h的头文件

### DISCUSSION THREAD

**Discussion by xqyjlj** [xmake-io/xmake#3041] [Q&A] [answered]
我在计划将xmake移植到一个可裁剪的操作系统上，原有的操作系统是采用kconfig生成config.h的形式，在[对于难以修改为option的kconfig，有什么办法能直接使用或其它解决办法？](https://github.com/xmake-io/xmake/discussions/2046)中，我了解到可以使用option代替kconfig，在option中，我采用以下方式：

```lua
option("USING_USER_MAIN")
    set_default(true)
    set_showmenu(true)
    add_defines("USING_USER_MAIN")
```

此方式是在编译时自动加上 -DUSING_USER_MAIN，在少数宏时比较好使，但是如果有大量宏时，不仅难以观察，过长的命令行也会引起编译报错。

之后我又留意到 [set_configvar](https://xmake.io/#/zh-cn/manual/project_target?id=targetset_configvar) 参数，但是此方式就需要维护两份文件了，一份option，一份模板文件。那么是否存在一种方式，能够直接将option转为头文件呢

**Comment by waruqi** [comment, marked as answer]
add_configfiles + set_configvar 或者 https://xmake.io/#/zh-cn/manual/helper_interfaces 

可以不用 option

**Reply by xqyjlj** [reply]
> 可以不用 option

不用option那就不能用menu对操作系统进行裁剪了，但是使用 add_configfiles + set_configvar 就必须要维护两份文件（xmake.lua+config.h.in）,我刚刚在源码中看了一下关于option的api，似乎可以用这种方式

```lua
before_build(function(target)
    import("core.project.project") 
    for targetname, target in pairs(project.targets()) do
        for optionname, option in pairs(project.options()) do
            print(optionname)
            print(option:value())
        end
    end
end)   
```

在这里收集option的相关信息，然后输出到config.h中。

**Reply by waruqi** [reply]
维护 config.h.in 是推荐做法，那个只是模版文件，set_configvar 也是可以设置到 option 里的。

你这么搞，每次 config.h 都会变，影响增量编译。

**Reply by xqyjlj** [reply]
翻阅了[仓库代码](https://github.com/xmake-io/xmake-repo/blob/695882ad9514681982fa483ce5b039e72165b89d/packages/l/libiconv/xmake.lua#L31-L36) 我想我已经找到合适的办法了，直接移植的时候用脚本快速生成一个 config.h.in 模板就行了
